programs/sunrise_beam/src/instructions/register_beam.rs

```diff
@@ -1,27 +1,16 @@
-use crate::state::{ControllerState, RegisterBeamInput};
-use crate::BeamProgramError;
 use anchor_lang::prelude::*;
 
-#[derive(Accounts)]
-#[instruction(input: RegisterBeamInput)]
-pub struct RegisterBeam<'info> {
-    #[account(mut, has_one = update_authority)]
-    pub state: Account<'info, ControllerState>,
-    #[account(mut)] // might have to pay for a resize
-    pub payer: Signer<'info>,
-    pub update_authority: Signer<'info>,
-    /// CHECK: The beam being registered is a valid account.
-    #[account(constraint = beam_state.key() == input.beam)]
-    pub beam_state: UncheckedAccount<'info>,
-    pub system_program: Program<'info, System>,
-}
+use crate::system::resize_state;
+use crate::{state::BeamDetails, RegisterBeam};
 
-pub fn handler(ctx: Context<RegisterBeam>, input: RegisterBeamInput) -> Result<()> {
+pub fn handler(ctx: Context<RegisterBeam>, beam_key: Pubkey) -> Result<()> {
     let state = &mut ctx.accounts.state;
 
-    // TODO: Reallocate space here instead of erroring?.
-    if state.push_allocation(input.into())?.is_none() {
-        return Err(BeamProgramError::WouldExceedBeamCapacity.into());
+    if state
```

Comment:
I think resizing and adding new beams should probably be separate instructions.

Reply:
Okay. Just to be clear, this instruction shouldn't do any size-checks or resize?

Reply:
There are two options I can think of for handling a resize:
- Always dynamically resize on-chain. Currently a resize usually just means an increase but on a decrease we can refund rent. We can figure if a resize is needed using any of the (current) fill-with-default method, checking the account's data len before an action, or storing an alloc_count. An off-chain resize would invalidate the first and third but be caught by the 2nd.

- Be optimistic when adding beams and leave it to the caller to handle resizes and ensure sanity.

Reply:
I would prefer the latter. In general, instructions should do as few things as possible. If you need to do two things, call two instructions. This puts more stran on the client but also adds flexibility, and makes things simpler, which reduces bugs or security flaws.